### DISCUSSION TITLE
dag.run & backfill causes OperationalError on MySQL and Postgres

### DISCUSSION THREAD

**Discussion by tvamsisai** [apache/airflow#29837] [Q&A]
For our integration tests setup we do something similar as mentioned in the testing guidance https://github.com/apache/airflow/blob/main/TESTING.rst#dag-testing 

However, since we needed to move to LocalExecutor from a SequentialExecutor with the migration to Airflow v2.2.5 from Airflow v1.10.15 we had to switch to MySQL or Postgres backend from SQLite.

Since this change we find that an arbitrary DagRun fails with
> Error
sqlalchemy.exc.OperationalError: (MySQLdb.OperationalError) (2013, 'Lost connection to MySQL server during query')
(Background on this error at: http://sqlalche.me/e/13/e3q8)
Stacktrace
/usr/local/lib/python3.8/site-packages/airflow/models/dag.py:2262: in run
    job.run()
/usr/local/lib/python3.8/site-packages/airflow/jobs/base_job.py:246: in run
    self._execute()
/usr/local/lib/python3.8/site-packages/airflow/utils/session.py:70: in wrapper
    return func(*args, session=session, **kwargs)
/usr/local/lib/python3.8/site-packages/airflow/jobs/backfill_job.py:820: in _execute
    self._execute_dagruns(
/usr/local/lib/python3.8/site-packages/airflow/utils/session.py:67: in wrapper
    return func(*args, **kwargs)
/usr/local/lib/python3.8/site-packages/airflow/jobs/backfill_job.py:726: in _execute_dagruns
    dag_run = self._get_dag_run(dagrun_info, dag, session=session)
/usr/local/lib/python3.8/site-packages/airflow/utils/session.py:67: in wrapper
    return func(*args, **kwargs)
/usr/local/lib/python3.8/site-packages/airflow/jobs/backfill_job.py:309: in _get_dag_run
    current_active_dag_count = dag.get_num_active_runs(external_trigger=False)
/usr/local/lib/python3.8/site-packages/airflow/utils/session.py:70: in wrapper
    return func(*args, session=session, **kwargs)
/usr/local/lib/python3.8/contextlib.py:120: in __exit__
    next(self.gen)
/usr/local/lib/python3.8/site-packages/airflow/utils/session.py:32: in create_session
    session.commit()
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py:1046: in commit
    self.transaction.commit()
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py:508: in commit
    t[1].commit()
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1762: in commit
    self._do_commit()
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1793: in _do_commit
    self.connection._commit_impl()
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:773: in _commit_impl
    self._handle_dbapi_exception(e, None, None, None, None)
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1510: in _handle_dbapi_exception
    util.raise_(
/usr/local/lib/python3.8/site-packages/sqlalchemy/util/compat.py:182: in raise_
    raise exception
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:771: in _commit_impl
    self.engine.dialect.do_commit(self.connection)

and each subsequent run of another Dag in the test runs fails with
> sqlalchemy.exc.InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (MySQLdb.OperationalError) (2013, 'Lost connection to MySQL server during query')
[SQL: INSERT INTO dag_run (dag_id, queued_at, execution_date, start_date, end_date, state, run_id, creating_job_id, external_trigger, run_type, conf, data_interval_start, data_interval_end, last_scheduling_decision, dag_hash) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)]
[parameters: ('test_connexity_export_csv_s3', None, datetime.datetime(2021, 7, 31, 0, 0), datetime.datetime(2023, 2, 28, 1, 46, 25, 911971), None, <TaskInstanceState.RUNNING: 'running'>, 'backfill__2021-07-31T00:00:00+00:00', 197, 0, <DagRunType.BACKFILL_JOB: 'backfill'>, b'\x80\x05}\x94.', datetime.datetime(2021, 7, 31, 0, 0), datetime.datetime(2021, 7, 31, 0, 0), None, None)]
(Background on this error at: http://sqlalche.me/e/13/e3q8) (Background on this error at: http://sqlalche.me/e/13/7s2a)

We have changed the backends between MySQL and Postgres with the same behaviour.

The `parallelism` and `concurrency` were tweaked down and up with no relief.
The resources of backend databases are uncapped on the docker containers that run on heavy machines.

The deployment of airflow and mysql/postgres is via a docker-compose not too different from
https://github.com/gtoonstra/etl-with-airflow/blob/master/docker-compose-LocalExecutor.yml

Any ideas on causes or issues will be appreciated. 🙏

**Comment by potiuk** [comment]
What the link describes is using DebugExecutor. Using Local Executor for testing is a bad idea and I don't think what you are trying to do is going to work if you use it.

**Reply by tvamsisai** [reply]
That's a shame. Using DebugExecutor would mean going back to SQLite and not having the tests running as fast.
I have tried kicking off DebugExecutor and Postgres to no avail, fails with the same errors.

Could you please provide more details on why LocalExecutor cannot work for this use case?

**Reply by potiuk** [reply]
I am not sure what is your case. So far you explained that you follow  https://github.com/apache/airflow/blob/main/TESTING.rst#dag-testing which is about DebugExecutor and instructions there describe the DebugExecutor testing, the instructions won't work for LocalExecutor, they are specifically describing how to use DebugExecutor to be able to run step-by-step debugging and testing your DAGs in single process (which the chapter is about):

> To ease and speed up the process of developing DAGs, you can use py:class:~airflow.executors.debug_executor.DebugExecutor, which is a single process executor for debugging purposes. Using this executor, you can run and debug DAGs from your IDE.

This is what I responded to when I said "Using Local Executor to do it is a bad idea".

Now - you mentioned using LocalExecutor which is quite a different case and you cannot do what is described in this chapter, so I am not sure what you are trying to do and how. For sure you are not trying to achieve what "dag-testing" chapter describes. It is quite impossible for me to answer your questions, because I have no idea what use case you are referring to and what you are trying to achieve. I think you should start with explaining what you want to do.
